### ISSUE TITLE
_lower_parfor_parallel function can mess up sequential_parfor_lowering flag disabling parallelization for the rest of the execution

### ISSUE THREAD

**Issue by Hardcode84** [numba/numba#5098]
## Reporting a bug

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Repro:
```
class DummyType(types.Opaque):
    pass

dummy_type = DummyType("my_dummy")
register_model(DummyType)(OpaqueModel)

class Dummy(object):
    pass

@typeof_impl.register(Dummy)
def typeof_Dummy(val, c):
    return dummy_type

@unbox(DummyType)
def unbox_index(typ, obj, c):
    return NativeValue(c.context.get_dummy_value())

@overload_method(DummyType, "method1", jit_options={"parallel":True})
def _get_method1(obj, arr, func):
    def _foo(obj, arr, func):
        def baz(a, f):
            c = a.copy()
            c[numpy.isinf(a)] = numpy.nan
            return f(c)

        length = len(arr)
        output_arr = numpy.empty(length, dtype=numpy.float64)
        for i in prange(length):
            output_arr[i] = baz(arr[i], func)
        for i in prange(length - 1):
            output_arr[i] += baz(arr[i], func)
        return output_arr
    return _foo

@njit
def bar(v):
    return v.mean()

@njit
def test1(d):
    return d.method1(numpy.array([[1.0,2.0,3.0],[4.0,5.0,6.0]]), bar)

print(numba.parfor.sequential_parfor_lowering)
test1(Dummy())
print(numba.parfor.sequential_parfor_lowering)
```
Prints
```
False
...some parfor warnings...
True
```

There are actually multiple issues:
* If `_create_gufunc_for_parfor_body` https://github.com/numba/numba/blob/master/numba/npyufunc/parfor.py#L239 throws an error `sequential_parfor_lowering` flag wont be restored
* The error itself looks like this
```
Traceback (most recent call last):
  File "/home/ibutygin/src/numba/numba/numba/errors.py", line 717, in new_error_context
    yield
  File "/home/ibutygin/src/numba/numba/numba/lowering.py", line 288, in lower_block
    self.lower_inst(inst)
  File "/home/ibutygin/src/numba/numba/numba/lowering.py", line 476, in lower_inst
    func(self, inst)
  File "/home/ibutygin/src/numba/numba/numba/npyufunc/parfor.py", line 241, in _lower_parfor_parallel
    bool(alias_map), index_var_typ, parfor.races)
  File "/home/ibutygin/src/numba/numba/numba/npyufunc/parfor.py", line 833, in _create_gufunc_for_parfor_body
    replace_var_with_array(races, loop_body, typemap, lowerer.fndesc.calltypes)
  File "/home/ibutygin/src/numba/numba/numba/npyufunc/parfor.py", line 1206, in replace_var_with_array
    replace_var_with_array_internal(vars, loop_body, typemap, calltypes)
  File "/home/ibutygin/src/numba/numba/numba/npyufunc/parfor.py", line 1203, in replace_var_with_array_internal
    block.body = replace_var_with_array_in_block(vars, block, typemap, calltypes)
  File "/home/ibutygin/src/numba/numba/numba/npyufunc/parfor.py", line 1189, in replace_var_with_array_in_block
    setitem_node = ir.SetItem(inst.target, const_var, inst.value, inst.loc)
  File "/home/ibutygin/src/numba/numba/numba/ir.py", line 573, in __init__
    assert isinstance(value, Var)
AssertionError
```
* Error get silently ignored if we use `@overload_method` (but it will fail compilation if we make `_foo` a free funtion)

Tested on `release0.47` branch and master (`40fb2d06859a782db9463ac6d7eae8405eb6a81d`)

**Comment by Hardcode84** [comment]
I can make a quick fix, wrapping `_create_gufunc_for_parfor_body` into `try..finally` but other issues are outside of my current competence

**Comment by stuartarchibald** [comment]
Whilst contrived, could this be used to reproduce and check the fix in #5114 ?
```python
from numba import njit, prange
import numba.parfors.parfor

from numba.core.compiler_machinery import register_pass, FunctionPass
from numba.core.compiler import CompilerBase, DefaultPassBuilder
from numba.core.typed_passes import IRLegalization

@register_pass(mutates_CFG=True, analysis_only=False)
class BreakParfors(FunctionPass):
    _name = "break_parfors"

    def __init__(self):
        FunctionPass.__init__(self)

    def run_pass(self, state):
        for blk in state.func_ir.blocks.values():
            for stmt in blk.body:
                if isinstance(stmt, numba.parfors.parfor.Parfor):
                    # races should be a set(), that list is iterable permits it to
                    # get through to the _create_gufunc_for_parfor_body routine
                    # at which point it needs to be a set so e.g. set.difference
                    # can be computed, this therefore creates an error in the
                    # right location
                    stmt.races = []
            return True


class BreakParforsCompiler(CompilerBase):

    def define_pipelines(self):
        pm = DefaultPassBuilder.define_nopython_pipeline(self.state)
        pm.add_pass_after(BreakParfors, IRLegalization)
        pm.finalize()
        return [pm]


@njit(parallel=True, pipeline_class=BreakParforsCompiler)
def foo():
    x = 1
    for _ in prange(1):
        x += 1
    return x


print(numba.parfors.parfor.sequential_parfor_lowering)
try:
    print(foo())
except Exception as e:
    print("Exception was raised:", e)
print(numba.parfors.parfor.sequential_parfor_lowering)
```
